lib/Doctrine/ORM/Mapping/DefaultTypedFieldMapper.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Mapping;
+
+use DateInterval;
+use DateTime;
+use DateTimeImmutable;
+use Doctrine\DBAL\Types\Type;
+use Doctrine\DBAL\Types\Types;
+use ReflectionEnum;
+use ReflectionNamedType;
+use ReflectionProperty;
+
+use function array_merge;
+use function assert;
+use function enum_exists;
+
+use const PHP_VERSION_ID;
+
+/** @psalm-type ScalarName = 'array'|'bool'|'float'|'int'|'string' */
+final class DefaultTypedFieldMapper implements TypedFieldMapper
+{
+    /** @var array<class-string|ScalarName, class-string<Type>|string> $typedFieldMappings */
+    protected $typedFieldMappings;
+
+    /** @param array<class-string|ScalarName, class-string<Type>|string> $typedFieldMappings */
+    public function __construct($typedFieldMappings = [])
+    {
+        $this->typedFieldMappings = $typedFieldMappings;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validateAndComplete(array $mapping, ReflectionProperty $field): array
+    {
+        $type = $field->getType();
+
+        if (
+            ! isset($mapping['type'])
+            && ($type instanceof ReflectionNamedType)
+        ) {
+            if (PHP_VERSION_ID >= 80100 && ! $type->isBuiltin() && enum_exists($type->getName())) {
+                $mapping['enumType'] = $type->getName();
+
+                $reflection = new ReflectionEnum($type->getName());
+                $type       = $reflection->getBackingType();
+
+                assert($type instanceof ReflectionNamedType);
+            }
+
+            $defaultTypedFieldMappings = [
+                DateInterval::class => Types::DATEINTERVAL,
+                DateTime::class => Types::DATETIME_MUTABLE,
+                DateTimeImmutable::class => Types::DATETIME_IMMUTABLE,
+                'array' => Types::JSON,
+                'bool' => Types::BOOLEAN,
+                'float' => Types::FLOAT,
+                'int' => Types::INTEGER,
+                'string' => Types::STRING,
+            ];
+
+            $typedFieldMappings = array_merge($defaultTypedFieldMappings, $this->typedFieldMappings);
```

Comment:
`array_merge()` is an expensive operation and `validateAndComplete()` will be called quite a few times. Given that we will always merge the same two arrays here, wouldn't it be better to perform this operation once and store the result?